tools/vp20compiler/main.c

```diff
@@ -540,9 +540,18 @@ void translate(const char* str)
                     // TODO: the index needs ajustment?
                     vsh_set_field(vsh_ins, FLD_CONST, reg.Index+96);
                 } else if (reg.File == PROGRAM_INPUT) {
+                    int in_reg;
+                    const char* name = _mesa_nv_vertex_input_register_name(reg.Index);
+                    for (in_reg = 0; _mesa_nv_vertex_hw_input_register_name(in_reg); in_reg++) {
+                        if (strcmp(name, _mesa_nv_vertex_hw_input_register_name(in_reg)) == 0) {
+                            break;
```

Comment:
This code checks, if a given input_register actually exists in hardware, right?

Reply:
The program in memory is in OpenGL format, but this code translates it to the hardware.

So this code retrieves the GL name by index using:
- `_mesa_nv_vertex_input_register_name[]`
and then loops over the hardware names using (note the `hw_`):
- `_mesa_nv_vertex_hw_input_register_name[]`
to find the respective hardware index (by matching the GL name and hw name in a sorted array).

Reply:
I see. I got confused because it looks similar to [this](https://github.com/XboxDev/nxdk/blob/629e55b196e72a23a5b0894b5021799a78e5a810/tools/vp20compiler/nvvertparse.c#L571) in `nvvertparse.c#L571`.  Case closed :)

Reply:
It *is* the same.

nvvertparse is the parser which creates structures from the program code, and at that time we have to check for parser errors / syntax errors. So that's when we have to report any problem with non-existing hardware registers (so we have to map it).

We could store the hardware index at that point (during parsing), but I decided against it because the mesa code expects the GL indices / names.
We might also want to provide extensions which adds virtual registers (which don't exist in hardware), so using hardware indices is probably worse than using the loosely defined GL indices which can be extended more easily.

Therefore, the `translate` function has to do it again (and using another interface, as it's in another file).

---


It's very possible that this will be refactored in the future. For now, I just wanted something that works. And this is suitable.